k2/csrc/fsa_equivalent.cc

```diff
@@ -62,64 +166,197 @@ bool IsRandEquivalent(const Fsa &a, const Fsa &b, std::size_t npath /*=100*/) {
+    // for non-top-sorted FSAs, but we prefer to decide this later as there's no
+    // such scenarios (input FSAs are not top-sorted) currently. If we finally
+    // find out that we don't need that version, we will remove flag
+    // `top_sorted` and add requirements as comments in the header file.
+    CHECK(top_sorted);
+    double sum_a =
+        ShortestDistance<Type>(a_compose_path, a_compose_weights.data());
+    double sum_b =
+        ShortestDistance<Type>(b_compose_path, b_compose_weights.data());
+    if (!DoubleApproxEqual(sum_a, sum_b)) return false;
+  }
+  return true;
+}
+
+// explicit instantiation here
+template bool IsRandEquivalent<kMaxWeight>(const Fsa &a, const float *a_weights,
+                                           const Fsa &b, const float *b_weights,
+                                           bool top_sorted, std::size_t npath);
+template bool IsRandEquivalent<kLogSumWeight>(
+    const Fsa &a, const float *a_weights, const Fsa &b, const float *b_weights,
+    bool top_sorted, std::size_t npath);
+
 bool RandomPath(const Fsa &a, Fsa *b,
                 std::vector<int32_t> *state_map /*=nullptr*/) {
-  if (IsEmpty(a) || b == nullptr) return false;
-  // we cannot do `connect` on `a` here to get a connected fsa
-  // as `state_map` will map to states in the connected fsa
-  // instead of in `a` if we do that.
-  if (!IsConnected(a)) return false;
+  return RandomPathHelper(a, b, false, state_map);
+}
 
-  int32_t num_states = a.NumStates();
-  std::vector<int32_t> state_map_b2a;
-  std::vector<int32_t> state_map_a2b(num_states, -1);
-  // `visited_arcs[i]` stores `arcs` leaving from state `i` in `b`
-  std::vector<std::unordered_set<Arc, ArcHash>> visited_arcs;
+bool RandomPathWithoutEpsilonArc(
+    const Fsa &a, Fsa *b, std::vector<int32_t> *state_map /*= nullptr*/) {
+  return RandomPathHelper(a, b, true, state_map);
+}
 
-  std::random_device rd;
-  std::mt19937 generator(rd());
-  std::uniform_int_distribution<int32_t> distribution(0);
+bool Intersect(const Fsa &a, const float *a_weights, const Fsa &b, Fsa *c,
+               std::vector<float> *c_weights,
+               std::vector<int32_t> *arc_map_a /*= nullptr*/,
+               std::vector<int32_t> *arc_map_b /*= nullptr*/) {
+  CHECK_NOTNULL(c);
+  CHECK_NOTNULL(c_weights);
+  c->arc_indexes.clear();
+  c->arcs.clear();
+  c_weights->clear();
+  if (arc_map_a != nullptr) arc_map_a->clear();
+  if (arc_map_b != nullptr) arc_map_b->clear();
 
-  int32_t num_visited_arcs = 0;
-  int32_t num_visited_state = 0;
-  int32_t state = 0;
-  int32_t final_state = num_states - 1;
-  while (true) {
-    if (state_map_a2b[state] == -1) {
-      state_map_a2b[state] = num_visited_state;
-      state_map_b2a.push_back(state);
-      visited_arcs.emplace_back(std::unordered_set<Arc, ArcHash>());
-      ++num_visited_state;
+  if (IsEmpty(a) || IsEmpty(b)) return true;
+  if (!IsArcSorted(a) || !IsArcSorted(b)) return false;
```

Comment:
I think we should be crashing rather than returning false if these conditions are not met.  Could be CHECK statements.

Reply:
Oh, yes, you're right, Thanks. I also changed the return type from `bool` to `void`. You can merge it if everything is fine.